internal/enricher/geolocation.go

```diff
@@ -0,0 +1,259 @@
+		// if the context section is missing on the event
+		// set it with default as map[string]interface{}
+		if _, ok := event["context"]; !ok {
+			event["context"] = map[string]interface{}{}
+		}
+
+		// if the context is other than map[string]interface{}, continue
+		context, ok := event["context"].(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		// if the `geo` key already present on the event, continue
+		if _, ok := context["geo"]; ok {
+			continue
+		}
+
+		contextIP, _ := context["ip"].(string)
+		ip := firstNonBlankValue(contextIP, request.RequestIP) // ip could still be blank given both are blanks
+
+		errType := ""
+		rawGeo, err := e.fetcher.Locate(ip)
+		if err != nil {
+			if errors.Is(err, geolocation.ErrInvalidIP) {
+				// `emptyIP` even though it's invalid is treated differently
+				// to get better stats about how many non-empty values are coming in which are invalid.
+				if ip == "" {
+					errType = ERR_EMPTY_IP
+				} else {
+					errType = ERR_INVALID_IP
+				}
+			} else {
+				// empty / invalidIP's are not terminal errors but
+				// any error except that mean the database failed for lookup
+				errType = ERR_LOCATE_FAILED
+				enrichErrs = append(enrichErrs, fmt.Errorf("locating geolocation for ip: %w", err))
+			}
+		}
+
+		e.stats.NewTaggedStat(
+			"proc_geo_enricher_request",
+			stats.CountType,
+			stats.Tags{
+				"sourceId":    source.ID,
+				"workspaceId": source.WorkspaceID,
+				"sourceType":  source.SourceDefinition.Type,
+				"error":       errType,
+			}).Increment()
+
+		// Set the empty data on the context nonetheless
+		context["geo"] = extractGeolocationData(ip, rawGeo)
+	}
+
+	return errors.Join(enrichErrs...)
+}
+
+// firstNonBlankValue iterates over the array and returns first
+// non blank aka non empty value from the list
+func firstNonBlankValue(vals ...string) string {
+	if len(vals) == 0 {
+		return ""
+	}
+
+	for _, val := range vals {
+		if val != "" {
+			return val
+		}
+	}
+	// return end most value
+	return vals[len(vals)-1]
+}
+
+func (e *geoEnricher) Close() error {
+	e.logger.Info("closing the geolocation enricher")
+
+	if err := e.fetcher.Close(); err != nil {
+		return fmt.Errorf("closing the geo enricher: %w", err)
+	}
+	return nil
+}
+
+// downloadMaxmindDB downloads database file from upstream s3 and stores it in
+// a specified location
+func downloadMaxmindDB(ctx context.Context, conf *config.Config, log logger.Logger) (string, error) {
+	var (
+		dbKey   = conf.GetString("Geolocation.db.key", "geolite2City.mmdb")
+		bucket  = conf.GetString("Geolocation.db.bucket", "rudderstack-geolocation")
+		region  = conf.GetString("Geolocation.db.bucket.region", "us-east-1")
+		timeout = conf.GetInt("Geolocation.db.download.timeout", 100)
+	)
+
+	log.Infof("downloading new geolocation db from key: %s", dbKey)
+	var (
+		tmpDirPath = strings.TrimSuffix(conf.GetString("RUDDER_TMPDIR", "."), "/")
+		baseDIR    = fmt.Sprintf("%s/geolocation", tmpDirPath)
```

Comment:
I believe this can be simplified
```suggestion
		baseDIR  = path.Join(conf.GetString("RUDDER_TMPDIR", "."), "geolocation")
```